compileScala throws java.lang.UnsupportedOperationException since 8.0-milestone-5

<!--- 
Please use our bug report template to report problems with something that has never worked.  
Regressions reports are greatly appreciated during our RC phase and before a final release.
-->
The following <code>build.gradle.kts</code>
<pre>
plugins {
  scala
}

repositories {
  mavenCentral()
}

dependencies {
  implementation("org.scala-lang:scala3-library_3:3.2.1")
}

tasks.withType<ScalaCompile> {
  scalaCompileOptions.additionalParameters = mutableListOf("-feature") // OK
  scalaCompileOptions.additionalParameters = listOf("-feature") // runtime java.lang.UnsupportedOperationException
}
</pre>
compiles but fails to <code>compileScala</code>.

This happens since <code>8.0-milestone-5</code>, <code>8.0-milestone-4</code> and <code>7.6</code> work fine.

There have been made some changes to <code>AbstractScalaCompile.java</code> on Dec 1, 2022. This may explain why milestone 4 works and milestone 5 fails.

This [scalaproject.zip](https://github.com/gradle/gradle/files/10242701/scalaproject.zip) contains the Scala project which does not work.
<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->
<pre>$ gradlew compileScala</pre>

should run without errors.
### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->

<pre>$ gradlew compileScala
Task :compileScala FAILED
FAILURE: Build failed with an exception.
What went wrong:
Execution failed for task ':compileScala'.
 java.lang.UnsupportedOperationException (no error message)
</pre>
The build script compiles but fails to <code>compileScala</code>.
### Context
<!--- Which version did this stop working with? Which version do you know it works? -->
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Fails with <code>8.0-milestone-5</code>, works with <code>8.0-milestone-4</code> and <code>7.6</code>.
### Steps to Reproduce 
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->
Unpack the attached <code>scalaproject.zip</code> and run <code>gradlew compileScala</code>.
### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:

--- a/subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc
+++ b/subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc
@@ -585,6 +585,12 @@ In case no toolchains are explicitly configured, the toolchain corresponding to 
 
 Similarly, tasks from the Groovy and Scala plugins also rely on toolchains to determine on which JVM they are executed.
 
+==== Scala compile options `getAdditionalParameters()` returns immutable list
+
+The list of additional parameters for Scala compilation retrieved from the `ScalaCompileOptions` or `BaseScalaCompileOptions` is now immutable.
+
+Previously, the mutability of the list depended on the last assigned value.
+
 [[changes_7.6]]
 == Upgrading from 7.5 and earlier
 

--- a/subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/ScalaCompileIntegrationTest.groovy
+++ b/subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/ScalaCompileIntegrationTest.groovy
@@ -72,4 +72,28 @@ class Person {
         }
     }
 
+    def "can assign #value to additional parameters"() {
+        file("src/main/scala/ScalaHall.scala") << "class ScalaHall(name: String)"
+
+        buildFile << """
+            tasks.withType(ScalaCompile) {
+              scalaCompileOptions.additionalParameters = $expression
+            }
+        """
+
+        when:
+        run ":compileScala"
+
+        then:
+        executedAndNotSkipped(":compileScala")
+        outputDoesNotContain("[Warn]")
+        JavaVersion.forClass(scalaClassFile("ScalaHall.class").bytes) == JavaVersion.VERSION_1_8
+
+        where:
+        value            | expression
+        "null"           | "null"
+        "mutable list"   | "[]"
+        "immutable list" | "[].asImmutable()"
+    }
+
 }

--- a/subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/ScalaCompileOptionsConfigurer.java
+++ b/subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/ScalaCompileOptionsConfigurer.java
@@ -16,6 +16,7 @@
 
 package org.gradle.api.tasks.scala.internal;
 
+import com.google.common.collect.ImmutableList;
 import org.gradle.api.tasks.scala.ScalaCompileOptions;
 import org.gradle.jvm.toolchain.JavaInstallationMetadata;
 import org.gradle.jvm.toolchain.internal.JavaToolchain;
@@ -77,7 +78,7 @@ public class ScalaCompileOptionsConfigurer {
         if (additionalParameters == null) {
             scalaCompileOptions.setAdditionalParameters(Collections.singletonList(targetParameter));
         } else {
-            additionalParameters.add(targetParameter);
+            scalaCompileOptions.setAdditionalParameters(new ImmutableList.Builder<String>().addAll(additionalParameters).add(targetParameter).build());
         }
     }
 

--- a/subprojects/scala/src/main/java/org/gradle/language/scala/tasks/BaseScalaCompileOptions.java
+++ b/subprojects/scala/src/main/java/org/gradle/language/scala/tasks/BaseScalaCompileOptions.java
@@ -16,6 +16,7 @@
 
 package org.gradle.language.scala.tasks;
 
+import com.google.common.collect.ImmutableList;
 import org.gradle.api.Incubating;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.provider.Property;
@@ -164,14 +165,18 @@ public abstract class BaseScalaCompileOptions extends AbstractOptions {
     /**
      * Additional parameters passed to the compiler.
      * Each parameter must start with '-'.
+     *
+     * @return The immutable list of additional parameters.
      */
-    @Nullable @Optional @Input
+    @Nullable
+    @Optional
+    @Input
     public List<String> getAdditionalParameters() {
         return additionalParameters;
     }
 
     public void setAdditionalParameters(@Nullable List<String> additionalParameters) {
-        this.additionalParameters = additionalParameters;
+        this.additionalParameters = additionalParameters == null ? null : ImmutableList.copyOf(additionalParameters);
     }
 
     /**

--- a/subprojects/scala/src/test/groovy/org/gradle/scala/compile/internal/ScalaCompileOptionsConfigurerTest.groovy
+++ b/subprojects/scala/src/test/groovy/org/gradle/scala/compile/internal/ScalaCompileOptionsConfigurerTest.groovy
@@ -30,6 +30,7 @@ class ScalaCompileOptionsConfigurerTest extends Specification {
     def 'using Java #toolchain and Scala #scalaLibraryVersion results in #expectedTarget'() {
         given:
         ScalaCompileOptions scalaCompileOptions = TestUtil.newInstance(ScalaCompileOptions)
+        scalaCompileOptions.additionalParameters = ["-some-other-flag"].asImmutable()
         def isScala3 = scalaLibraryVersion.startsWith("3.")
         File scalaLibrary = new File(isScala3 ? "scala3-library_3-${scalaLibraryVersion}.jar" : "scala-library-${scalaLibraryVersion}.jar")
         Set<File> classpath = [scalaLibrary]

